Support package@version syntax for dotnet workload search/update/install version Fixes #42367

Forgind opened the pull request:
Fixes #42367

Using `dotnet workload search version`, you should now be able to provide package@version as an argument, and it will find the highest-valued workload version that has that package at that version.

Sample when no such workload version was found:
![image](https://github.com/user-attachments/assets/fbf445a3-9164-4780-9bf0-8f98fb309b16)

When 3 were found:
![image](https://github.com/user-attachments/assets/95a2ae26-47b7-4417-aff9-c443e285738c)

(Note that the other two were 9.0.200.1 and 9.0.200, so this is the highest version number.)

There are three total uses of dotnet workload search version that previously existed:
Without any argument, it lists out the most recent SDKs (in the current feature band):
> dotnet workload search version --format json

![image](https://github.com/user-attachments/assets/bbe44f1c-3eaa-4b71-8fe5-db31cd381754)

When provided with a workload version, it lists out all the workloads in that workload version:
> dotnet workload search version 9.0.201

![image](https://github.com/user-attachments/assets/73e4ec4b-5278-4049-951e-ccabf71e8eac)

(Note that this was a manually created 9.0.201 version, so those versions do not align with a real one.)

And as mentioned earlier, when one or more workloads along with versions (concatenated with '@'), it finds a workload version that includes all of the provided workloads at the associated versions.
> dotnet workload search version maui@9.0.0-rc.1.24453.9 ios@17.5.9270-net9-rc1

![image](https://github.com/user-attachments/assets/8ee7936b-c4b4-43d9-82a8-40913f67102d)

This PR also dramatically changes the help text either when you use `dotnet workload search version -?` or when you mistype a command. It now looks like this:
![image](https://github.com/user-attachments/assets/232af696-458d-4813-a265-d69ef99ffbf4)

Forgind committed:
Attempt to support package@version syntax

Using `dotnet workload search version`, you should now be able to provide package@version as an argument, and it will find the highest-valued set that has that package at that version

Forgind committed:
Support multiple workloads

Forgind committed:
Merge branch 'release/9.0.2xx' of https://github.com/dotnet/sdk into search-version-with-component

Forgind commented on src/Cli/dotnet/commands/dotnet-workload/search/WorkloadSearchVersionsCommand.cs:
```diff
@@ -100,6 +90,30 @@ public override int Execute()
                     Reporter.WriteLine(string.Join('\n', versions));
                 }
             }
+            else if (_workloadVersion.Contains('@'))
```
Because of this usage, I don't think _workloadVersion is really a good name anymore. I'm leaving it for now for two reasons:
1. We might switch to --component (or something) later, in which case changing it is unnecessary.
2. Having two possible meanings means I'm not quickly coming up with a good name.

We should change it (if we don't switch to --component) before merging.

dsplaisted replied:
Yeah, it seems we are kind of mixing up different commands.  It looks like currently `dotnet workload search version` gives you a list of available workload set versions for the current feature band, `dotnet workload search version 9.0.100` gives you the manifest versions for the specified workload set version (same as the rollback file basically), and `dotnet workload search version maui@9.0.0` gives you the workload set versions with the specified manifest version that contains the specified workload.

@baronfel What do you think?  Should we be designing these commands differently?

Forgind committed:
Switch to using workload instead of workload manifest

Forgind committed:
Support install + update via @ syntax

Forgind committed:
Add test

Forgind committed:
Add workload search tests

dsplaisted commented on src/Cli/dotnet/commands/InstallingWorkloadCommand.cs:
```diff
@@ -242,6 +243,20 @@ protected void UpdateWorkloadManifests(WorkloadHistoryRecorder recorder, ITransa
                 return;
             }
 
+            if (resolvedWorkloadSetVersion?.Contains('@') == true)
```
For `install` and `update`, it's the workload ID argument that should be used to specify the version.  For example, `dotnet workload install ios@17.5.9231-net9-p7 tvos@17.5.9232-net9-p7`.  Note that the different workloads are separated by a space here, not a semicolon.

Forgind replied:
This doesn't align with the guidance marcpopMSFT provided:
> We decided to just use the existing --version or workloadVersion in the global.json for this and look for the @ symbol.

https://github.com/dotnet/sdk/issues/42367#issuecomment-2489867789

Did I miss a meeting on this in which we changed the plan?

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/search/WorkloadSearchVersionsCommand.cs:
```diff
@@ -123,5 +125,47 @@ public override int Execute()
+            }
+            catch (NuGetPackageNotFoundException)
+            {
+                Microsoft.DotNet.Cli.Utils.Reporter.Error.WriteLine(string.Format(LocalizableStrings.NoWorkloadVersionsFound, featureBand));
```
I would probably not output a message from this method if no packages were found, and leave it to callers to handle that.  That's the behavior I would be more likely to expect when calling a method to get the versions, and it would let us have more specific messages for each case (ie "No workload sets were found for the current feature band" vs "No workload sets were found with the specified workload version(s)").

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/search/WorkloadSearchVersionsCommand.cs:
```diff
@@ -123,5 +125,47 @@ public override int Execute()
+
+        public string FindBestWorkloadSetFromComponents()
+        {
+            var versions = GetVersions(0);
```
Does 0 here mean to get all the versions?  It would probably be good to clarify this.

Forgind replied:
Yes. Do you think a comment would suffice, or how else would you recommend making it clearer?

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/search/WorkloadSearchVersionsCommand.cs:
```diff
@@ -123,5 +125,47 @@ public override int Execute()
+            });
+
+            // Since these are ordered by version (descending), the first is the highest version
+            return versions.FirstOrDefault(version =>
```
When doing a search, we should display all of the matching workload set versions.  When doing an update or install, we should take the highest workload set version that matches.

dsplaisted commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadManifest.cs:
```diff
@@ -10,6 +10,8 @@ namespace Microsoft.NET.Sdk.WorkloadManifestReader
     /// </summary>
     public class WorkloadManifest
     {
+        public WorkloadManifest(string id) : this(id, new FXVersion(7, 3, 5), null, string.Empty, [], [], []) { }
+
```
This looks like it's just for tests, but that wouldn't be obvious if you were trying to create a `WorkloadManifest` in the product code.  Could you do one of the following?

- Create a helper method in the tests that creates a WorkloadManifest with some default values given an ID?
- Instead of a constructor, have a static `CreateForTests` factory method on the `WorkloadManifest` class so that it's clear what the purpose of the method is.

Forgind replied:
Good point. I like the idea of the static CreateForTests factory method; will implement that.

dsplaisted commented on src/Cli/dotnet/commands/InstallingWorkloadCommand.cs:
```diff
@@ -242,6 +243,20 @@ protected void UpdateWorkloadManifests(WorkloadHistoryRecorder recorder, ITransa
+            if (resolvedWorkloadSetVersion?.Contains('@') == true)
+            {
+                var searchVersionsCommand = new WorkloadSearchVersionsCommand(
+                     Parser.Instance.Parse("dotnet workload search version " + resolvedWorkloadSetVersion),
```
If we have to create an instance of another command class, I think it's better if we don't generate a command line and parse it.  Rather, we should have constructor parameters or class properties that let us pass in the data we need.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/search/WorkloadSearchVersionsCommand.cs:
```diff
@@ -123,5 +125,47 @@ public override int Execute()
+            var packageNamesAndVersions = workloadVersions.Select(version =>
+            {
+                var split = version.Split('@');
+                return (new ManifestId(_resolver.GetManifestFromWorkload(new WorkloadId(split[0])).Id), new ManifestVersion(split[1]));
```
`GetManifestFromWorkload` has to actually download and extract the NuGet package.  We might want to add some parallelism to those downloads, but it's probably a bit tricky to do so safely, so I would leave it as it is for now.

dsplaisted reviewed (commented):
Thanks, this is good progress.

Forgind commented on src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:
```diff
@@ -567,7 +567,8 @@ await Task.WhenAll(
                 {
-                    return stableVersions.OrderByDescending(r => r.package.Identity.Version).Take(numberOfResults);
+                    var results = stableVersions.OrderByDescending(r => r.package.Identity.Version);
+                    return numberOfResults > 0 ? results.Take(numberOfResults) : results;
```
I don't think I made this very clear. I initially made this method to return a certain number of workload sets, but once I added support for @ syntax, I needed a way to indicate 'get all versions', and 0 is sometimes used as a special indicator for 'all'

How would you suggest making that clearer? Would a comment suffice?

dsplaisted replied:
OK, that makes sense, I was thinking `numberOfResults` was the actual number of results returned by the query, not the requested number.  So it makes sense now.

Forgind committed:
Most comments

Forgind committed:
Feedback from design meeting

Forgind commented:
@dsplaisted I think I addressed all your feedback and the feedback from the design meeting earlier except for tweaking the help text and making it clearer what the 0 means 🙂 I'll fix the help text and add a comment tomorrow.

Forgind commented:
> Other variants are:
> 
> * `dotnet add <project> package <name> --version <ver>`
> * `dotnet new install <name>::<ver>`
> 
> `@` syntax is certainly much better than `::` syntax chosen for dotnet-new template install command. e.g. https://www.nuget.org/packages/BenchmarkDotNet.Templates shows `dotnet new install BenchmarkDotNet.Templates::0.14.0`. It would be nice to streamline these by deprecating `::` delimiter in favor of `@` and support `@` in dotnet add package as well.

I don't think this is really connected to this PR, but improving our CLI is a direction we'd like to go, and I agree that @ seems like a better separator for this sort of thing than ::

I'm going to turn this into a separate issue for triage if that's ok with you!

Forgind committed:
More feedback

Forgind committed:
Merge branch 'release/9.0.2xx' of https://github.com/dotnet/sdk into search-version-with-component

Forgind committed:
Small improvements

dsplaisted commented on src/Cli/dotnet/commands/InstallingWorkloadCommand.cs:
```diff
@@ -441,9 +468,10 @@ protected IEnumerable<WorkloadId> WriteSDKInstallRecordsForVSWorkloads(IEnumerab
+        public static readonly CliOption<IEnumerable<string>> WorkloadSetVersionOption = new("--version")
         {
-            Description = Strings.WorkloadSetVersionOptionDescription
+            Description = Strings.WorkloadSetVersionOptionDescription,
```
The help text you added is great, but I think the option text should also be updated to indicate it can be used in two ways.  Something like "A workload version to display, or a workload and its version joined by the '@' character.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs:
```diff
@@ -1183,5 +1183,10 @@ void IInstaller.UpdateInstallMode(SdkFeatureBand sdkFeatureBand, bool? newMode)
             Reporter.WriteLine(string.Format(LocalizableStrings.UpdatedWorkloadMode, newModeString));
         }
+
+        // This method should never be called for this kind of installer. It is challenging to get this information from an MSI
+        // and totally unnecessary as the information is identical from a file-based installer. It was added to IInstaller only
+        // to facilitate testing. As a consequence, it does not need to be implemented.
+        public WorkloadSet GetWorkloadSetContents(string workloadVersion) => throw new NotImplementedException();
```
We actually do this already for MSIs.  Check out the `ExtractManifestAsync` implementation.  In fact, you may be able to implement `GetWorkloadSetContents` in terms of that, possibly only in test code instead of in the product.

Forgind replied:
I don't know if I'm understanding this correctly, but it looks like it extracts information by installing the MSI?

https://github.com/dotnet/sdk/blob/0113afe0e7777a01a21cc7a985fad9075734a660/src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs#L792

That seems like it would change what you have installed, which wouldn't necessarily be what we'd want.

dsplaisted replied:
It does an "admin install", which means it doesn't actually install it, it just extracts the MSI to the target path.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/search/LocalizableStrings.resx:
```diff
@@ -136,7 +136,36 @@
+           microsoft.net.workload.mono.toolchain.net7         9.0.100-rc.1               9.0.0-rc.1.24431.7
+           microsoft.net.workload.mono.toolchain.net8         9.0.100-rc.1               9.0.0-rc.1.24431.7
+           microsoft.net.sdk.aspire                           8.0.100                    8.2.0
+      3. If a set of workloads are provided along with their versions, it outputs workload versions that match the provided versions. Takes the --take option to specify how many to provide and --format to alter the format.
```
```suggestion
      3. If one or more workloads are provided along with their versions (by joining them with the '@' character), it outputs workload versions that match the provided versions. Takes the --take option to specify how many to provide and --format to alter the format.
```

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/search/WorkloadSearchVersionsCommand.cs:
```diff
@@ -102,7 +123,7 @@ public override int Execute()
             else
             {
-                var workloadSet = _installer.GetWorkloadSetContents(_workloadVersion);
+                var workloadSet = _installer.GetWorkloadSetContents(_workloadVersion.Last());
```
Probably it should be an error if you specify two workload set versions, instead of just taking the last one.

Forgind replied:
It is! (Via the command parser rather than the command). Last, First, and Single should all be equivalent here. I only chose Last because I'd initially been thinking about not making it an error and decided on taking the last one, then decided that wasn't the best design and ended up at the same place as you.

Forgind replied:
I changed it to Single for clarity

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/search/WorkloadSearchVersionsCommand.cs:
```diff
@@ -123,5 +144,53 @@ public override int Execute()
+                return null;
+            }
+
+            var packageNamesAndVersions = workloadVersions.Select(version =>
```
Nit: better name

```suggestion
            var manifestIdsAndVersions = workloadVersions.Select(version =>
```

dsplaisted reviewed (approved):
Looks good, thanks!

Forgind commented on src/Cli/dotnet/commands/InstallingWorkloadCommand.cs:
```diff
@@ -214,7 +209,39 @@ protected void UpdateWorkloadManifests(WorkloadHistoryRecorder recorder, ITransa
+            }
+            else if (SpecifiedWorkloadSetVersionOnCommandLine)
+            {
+                resolvedWorkloadSetVersion = _workloadSetVersionFromCommandLine.Single();
```
Yep! I noticed that as part of my testing and added a validator to the command parser:
https://github.com/dotnet/sdk/pull/45156/files#diff-77f29f3a81e78911bdd0a0c1415d295295f17b18a8dbc622f89e920e243554daL63

Forgind replied:
Note that for this case, I specifically requested exactly 1 workload set, so there should never be more than 1, and we already checked for 0.

Forgind committed:
Feedback

Forgind committed:
Change 'multiple components' style in tests

Forgind commented:
> > This PR also dramatically changes the help text either when you use dotnet workload search version -? or when you mistype a command.
> 
> Does `dotnet workload search --help` (without `version`) also show the 29-line description of `dotnet workload search version`? IMO it shouldn't output so much text. Instead of putting all of that into the description, I think it would better as a separate help section.

I was surprised by this, but this appears to be intentional...there's a WriteSubcommand function when printing help text that is intended to tell you about all the subcommands, and that's exactly what this does.

On the one hand, this does feel too verbose to me. Perhaps we should cut it for dotnet workload search version. But on the other hand, what exactly that subcommand does is a bit confusing in the first place, hence the unusually complete description; I'm not sure I could craft a short description that clearly explains what all three possibilities are.

@baronfel, do you think we should keep the current (verbose but clear) text or try to shorten it? If the latter, I think I'd just put in a "if you're writing help for dotnet workload search, use this instead of the normal subcommand description."

Picture to give you an idea of what it looks like:
![image](https://github.com/user-attachments/assets/143645bc-cf4d-493c-9d3c-21e5cec9393f)

Forgind commented:
NTS: Something like https://github.com/dotnet/sdk/blob/9c97bb9b8b747d19fdca2f7db2c612de3fa6f317/src/Cli/dotnet/Parser.cs#L240 should work

Forgind committed:
ios --> android

Apparently ios isn't found on windows x64?

Forgind committed:
Tweak test to use a mock resolver

Forgind commented:
New version:
![image](https://github.com/user-attachments/assets/bb43b772-4ffb-4247-a988-9f99bc0b846a)

Forgind committed:
Shorten description for version as a subcommand

dsplaisted commented:
@forgind Is this ready to merge, or are you waiting for more feedback or for the CI issues to be fixed?

baronfel commented on src/Cli/dotnet/CommonLocalizableStrings.resx:
```diff
@@ -715,6 +715,10 @@ The default is 'false.' However, when targeting .NET 7 or lower, the default is
   </data>
+  <data name="ShortWorkloadSearchVersionDescription" xml:space="preserve">
+    <value>Search for available workload versions or what comprises a workload version. Use 'dotnet workload search version --help' for more information.</value>
+    <comment>dotnet workload search version --help should not be localized.</comment>
```
nit: use the [locking syntax](https://github.com/dotnet/sdk/blob/main/documentation/project-docs/Localization.md#locking-translations) for pinned translation strings so that the Loc team's tooling helps prevent unintentional translation.

Forgind committed:
Use {Locked}
